text/0000-symbol-name-mangling-v2.md

```diff
@@ -0,0 +1,799 @@
+
+
+```
+
+### Generic Functions
+
+Each monomorphization of a generic function has its own symbol name. The monomorphizations are disambiguated by the list of concrete generic arguments. These arguments are listed as a suffix, starting with `I`, after the name they belong to. So the instance
+
+```rust
+std::mem::align_of::<f64>
+```
+
+would be mangled to
+
+```
+_RN12std_a1b2c3d43mem8align_ofVIdEE
+                               ^^^
+                               |||
+     start of argument list ---+|+--- end of argument list
+                                |
+                               f64
+```
+
+where `I` starts the list of arguments, `d` designates `f64` and `E` ends the argument list. As we can see, we need to be able to represent all kinds of types that can be part of such an argument list. (In the future, when const generics get added to the language, we might also need to represent *values*) These kinds of types are:
+
+ - basic types (`char`, `()`, `str`, `!`, `i8`, `i16`, ...)
+ - reference and pointers types, shared and `mut`
+ - tuples
+ - arrays, with and without fixed size (e.g. `[u8]`, `[u8; 17]`, or as part of a slice type `&[char]`)
+ - structs, enums, closures, and other named types, possibly with their own set of type arguments
+ - function types such as `fn(&i32) -> u16`
+
+Basic types are all encoded via a single lower-case letter, like in the Itanium scheme. Named types are encoded as their absolute path (including arguments) like is done for function symbols. Composites like references, tuples, and function types all follow a simple grammar given in the reference-level explanation below. Here are some example manglings to get a general feel of what they look like:
+
+ - `std::mem::align_of::<usize>`: `_RN12std_a1b2c3d43mem8align_ofVIjEE`
+ - `std::mem::align_of::<&char>`: `_RN12std_a1b2c3d43mem8align_ofVIRcEE`
+ - `std::mem::align_of::<std::mem::Discriminant>`: `_RN12std_a1b2c3d43mem8align_ofVIN12std_a1b2c3d43mem12DiscriminantEEE`
+ - `std::mem::align_of::<&mut (&str,())>`: `_RN12std_a1b2c3d43mem8align_ofVIWTRrvEEE`
+
+There's one more thing we have to take into account for generic functions: The compiler may produce "crate-local" copies of a monomorphization. That is, if there is a function `foo<T>` which gets used as `foo<u32>` in two different crates, the compiler (depending on the optimization level) might generate two distinct functions at the LLVM IR level, each with it's own symbol. In order to support this without running into conflicts, symbol names for monomorphizations must include the id of the crate they are instantiated for. This scheme does this by appending an `<crate-id>` suffix to the symbol. So for example the mangling for `std::mem::align_of::<usize>` would actually look like this:
+
+```
+_RN12std_a1b2c3d43mem8align_ofVIjEE12foo_a1b2c3d4 (for crate "foo[a1b2c3d4])
+_RN12std_a1b2c3d43mem8align_ofVIjEE12bar_11223344 (for crate "bar[11223344])
+```
+
+
+### Closures and Closure Environments
```

Comment:
Make sure that whatever we do for closures it can handle generic closures if we ever add those. =)

Reply:
On that note: we should make closures not have `{{` and also probably add a line number or types to the closure name. It gets pretty frustrating seeing `{{closure}}` everywhere in backtraces ...

I made a pr for this once. It made it so you could set a breakpoint on the closure or print its address in a debugger, etc. because you can’t do that now in gdb afaik due to `{{`

Reply:
The `{{...}}` has nothing to do with mangling or debuginfo, it's an implementation detail of "def paths" (mostly used to approximate identity reuse for incremental), that's *accidentally leaking*, like some other things also are.

You'd literally see the same thing for impls (and methods in them) except we special-cased those in a very hacky way. Those hacks will hopefully go away with the move to an orthogonal "presentation" (encompasing both rustc diagnostics, and mangling), from the internal representation.

Reply:
@Centril Generic closures can be represented via a type-prefixed path like is done for inherent methods. So we should be good here.

Reply:
@michaelwoerister Great; Thank you!

Reply:
> I’ll note: your tone on irc was condescending and dismissive then and I find it dismissive now. It was this tone which also pushed me away from further work 9 months ago.

@m4b That wasn't my intention and I'm really sorry.

I don't recall the discussion and I can't find a branch/PR so I sadly can't say where I was wrong, but in retrospect I don't see why we wouldn't have accepted a PR that replaced the `{{closure}}` printing in symbol names/debuginfo with something more useful.

I screwed up, partly because I wasn't aware of how exactly C++11 does it (which, before this RFC, would've been been the way to go), but also because I kept hoping (since more than 2 years ago) we'd be doing what this RFC finally mandates us to do.

The «orthogonal "presentation"» bit in my comment above was also unclear because I forgot to point out explicitly that I'm referring to the implementation of this RFC.
That is, this RFC replaces "the compiler dumps its internals into symbol names in some arbitrary manner" with an encoding that each debugger can choose how to display and how to let the user interact with.

As for line/column numbers, I think they'd be useful (file name seems unnecessary because the symbol name contains the function the closure is in), although I worry a bit that it would break incremental, while debuginfo already has this information and [we can/should enable that more often](https://github.com/rust-lang/rust/pull/54459).

Reply:
@eddyb Ok thanks for clarifying; sorry I if my tone was unpleasant as well, I felt dismissed but I'm sure that wasn't your intention, thanks for meeting me halfway :)